### ISSUE TITLE
Misleading Error Message for Broken Code

### ISSUE THREAD

**Issue by xoebus** [elm-lang/elm-compiler#1206]
After a code change we came across an error message as we were chasing down the compiler errors that I don't think was as clear as it could be. (The error messages in 0.16 are otherwise fantastic. Thank you for working on them!).

We changed the `jobName` and `pipelineName` fields of the `Build` type alias below from `String` to `Maybe String`.

``` elm
import Html exposing (Html)
import Html.Attributes exposing (classList, href)


type alias Build =
  { id : Int
  , name : String
  , status : String
  , jobName : Maybe String
  , pipelineName : Maybe String
  , url : String
  }


viewBuildHeader : Build -> List Build -> Html
viewBuildHeader build history =
  Html.div []
    [ Html.text build.pipelineName
    , Html.text build.jobName
    , Html.ul [] (List.map (renderHistory build) history)
    ]

renderHistory : Build -> Build -> Html
renderHistory currentBuild build =
  Html.li
    [
      classList [
        (build.status, True),
        ("current", build.name == currentBuild.name)
      ]
    ]
    [ Html.a [href build.url] [ Html.text ("#" ++ build.name) ] ]
```

One of the compilation errors that this caused is below:

```
The argument to function `renderHistory` is causing a mismatch.

20|     , Html.ul [] (List.map (renderHistory build) history)

Function `renderHistory` is expecting the argument to be:

    { ..., jobName : Maybe String, pipelineName : Maybe String }

But it is:

    { ..., jobName : String, pipelineName : String }
```

This was confusing because it looks like the inferred type of the fields is taking precedence over the type annotation. I would expect the error message to be something saying that the argument to `Html.text` was `Maybe String` instead of `String`.

What do you think?

P.S. The code can be reduced further:

``` elm
import Html exposing (Html)

type alias Build =
  { jobName : Maybe String
  , pipelineName : Maybe String
  }

viewBuildHeader : Build -> Html
viewBuildHeader build =
  Html.div []
    [ Html.text build.pipelineName
    , Html.text build.jobName
    ]
```

```
The type annotation for `viewBuildHeader` does not match its definition.

10| viewBuildHeader : Build -> Html

The type annotation is saying:

    { jobName : Maybe String, pipelineName : Maybe String } -> Html

But I am inferring that the definition has this type:

    { jobName : String, pipelineName : String } -> Html
```

This error makes more sense as it directly states the mismatch between the annotation and inferred type. However, I normally make a change and then use the type system to help me find all the places that are affected and so here it would be nice to see the next change that I needed to make more clearly (the handling of `Maybe` in the call to `Html.text`).

**Comment by evancz** [comment]
Can you move this to https://github.com/elm-lang/error-message-catalog/issues where we track all confusing error messages? I think this is a duplicate of https://github.com/elm-lang/error-message-catalog/issues/28 in fact. If you can get this down to a better http://sscce.org than anything in there, it may be worth it to add as a comment.
